### ISSUE TITLE
Extended color theme

### ISSUE THREAD

**Issue by amoutonbrady** [tw-in-js/twind#11]
Hey, super cool project that’s probably going to change a lot the way I currently use tailwind in some of my projects. I was already quite into Oceanwind, but this is really next level.

I quickly tried and put it head to head, reproducing the welcome example of [https://play.tailwindcss.com/](https://play.tailwindcss.com/) into my favorite JS library and was blown away by how close it was… [In fact, i’d even say it’s 1:1](https://codesandbox.io/s/solid-twind-k0cn2).

There’s one thing, however, that might be cool to add and I didn’t see any mention in the documentation. The extended color palette introduced in tailwind 2. As far as I understand, the [default theme colors in twind](https://github.com/tw-in-js/twind/blob/main/src/twind/theme.ts#L92) is matching the [default theme in tailwind](https://github.com/tailwindlabs/tailwindcss/blob/master/stubs/defaultConfig.stub.js#L19), which only includes a small subset of colors. Here’s the full color list that was introduced in tailwind 2: [https://github.com/tailwindlabs/tailwindcss/blob/master/colors.js](https://github.com/tailwindlabs/tailwindcss/blob/master/colors.js).

For file size reason, I totally get that it shouldn’t be integrated within the default theme (that’s also probably why it wasn’t in tailwind by default, to trim down the compilation time), but I wonder if it could be possible to add like an extra exported object that contains those rules. e.g.:

```js
import { setup, colors } from 'twind'
// or import colors from 'twind/colors';

setup({
  theme: {
    extend: {
      colors: { cyan: colors.cyan }
    }
  }
})
```

The only problem I suppose is to get good tree shakeability from an object so that you only pay for the color you use… I’m not sure if that’d be possible. I guess you could also export each and every color separately:

```js
import { red, cyan, blue } from 'twind/colors';
// or import * as colors from 'twind/colors';

setup({
  theme: {
    extend: {
      colors: { red, cyan, blue }
    }
  }
})
```

That’s just a suggestion and I supposed one could also just copy and paste the the needed color from the original tailwind color palette as well.

Again, thanks for the library!

**Comment by lukejacksonn** [comment]
Hey man 👋 thanks for taking the time to try out twind and create this in depth issue! Very glad ytou have found the results satisfactory so far (we are aiming for a 1:1 mapping here) but you are totally right, we should have a way for developers to access all those new colors!

The colors file you linked to weighs `516 bytes` total and that is all the colors not just the new ones right? So extending the existing colors in the default theme with the news one here isn't as crazy as it might sound; note this is a fixed cost given that we don't generate a load of classes up front like Tailwind does.

In fact, here is what it would look like https://github.com/tw-in-js/twind/pull/12/files. It increases the bundle size by ~500 bytes. But requires no further thought, new export or external package etc.

Perhaps it is worth shipping this separately from core as you suggest. What are your thoughts @sastan?

**Comment by amoutonbrady** [comment]
Thanks for the quick reply, yes it’s the totality of the color palette. As you probably saw in the default theme, they just import a subset from that full color palette.

I reckon I didn’t do the exercise to see how much it would add to the bundle, 500 bytes doesn’t seem that much and is probably not worth the trouble of going ahead and fiddle with the setup function…

For what it’s worth, in my experience (I’ve used tailwind 2 on rather large projects professionally), I’ll always include the entirety of all of the colors & variants which makes the first compilation time slow but opens up to a lot of possibilities in dev. I wouldn’t be surprised that a lot of people would be in a similar scenario.

So if at the end of the day, most people would end up doing that:

```js
import * as colors from 'twind/colors';

setup({
  theme: {
    extend: { colors }
  }
})
```

You might as well add 500 bytes and remove that extra fiddling. I guess it depends on the final direction you want for the project. Am I correct in understanding that if you include the entirety, there would be no opt-out?

**Comment by lukejacksonn** [comment]
I have a similar experience when it comes to real life projects, the new colors are very useful. So building them in does seem sensible. However, as you rightly pointed out that would make it very hard (if not impossible) to opt out of.

It would not adversely affect our first compile time (because there isn't one!) and it would open up a lot of possibilities for developers. However, it would add 500 bytes to the module, making the module weigh **12kb gzipped** which is to too far away from other CSS-in-JS libraries.

Tough one.. will think about it a bit more and give anyone else the opportunity to chime in here!

**Comment by sastan** [comment]
> For what it’s worth, in my experience (I’ve used tailwind 2 on rather large projects professionally), I’ll always include the entirety of all of the colors & variants which makes the first compilation time slow but opens up to a lot of possibilities in dev. I wouldn’t be surprised that a lot of people would be in a similar scenario.

Understandable.

> So if at the end of the day, most people would end up doing that:
> 
> ```js
> import * as colors from 'twind/colors';
> 
> setup({
>   theme: {
>     extend: { colors }
>   }
> })
> ```
> 
> You might as well add 500 bytes and remove that extra fiddling. I guess it depends on the final direction you want for the project. Am I correct in understanding that if you include the entirety, there would be no opt-out?

I like the idea of adding `twind/colors` with named exports with all tailwind colors.

I'm will adjust the PR #12

**Comment by sastan** [comment]
@amoutonbrady I adjusted the PR #12 . Would that help?

**Comment by amoutonbrady** [comment]
Honestly, that looks exactly what I had in mind and looks much closer to what you’d end up doing according to tailwind 2 standards. You guys are awesome! That would totally help.

I didn’t investigate the whole code base and I’m not that familiar with monorepo, so nevermind me if that’s a silly suggestion, but if that PR go through, isn’t there a window for improvement in the [baseTheme](https://github.com/tw-in-js/twind/blob/main/src/twind/theme.ts#L92) to import colors from that new color file and have a single source of truth instead of maintaining it in both?

Thanks again, truly appreciate it!

**Comment by sastan** [comment]
Good idea. I'll try that.

**Comment by sastan** [comment]
Referencing one colors file and importing that in the default theme would increase the bundle size by 0.05kb: 

```
Before:       11.19 KB gzipped
After:         11.24 KB gzipped
```

That may not seem much but it may add up. I've added a test to ensure the default theme uses the correct/same colors.

**Comment by amoutonbrady** [comment]
Ah, that's awesome, thanks for the investigation. I totally get the file size concern, I try to promote that as well on my own hence why I suggested the `twind/colors` proposal in the first place to not lock people in.

Again, thanks for your reactivity!

**Comment by lukejacksonn** [comment]
Really like the result here. Great job everyone 🙌

**Comment by sastan** [comment]
@amoutonbrady This has been merged and released in v0.8.0.
